feat: support network IO traffic monitoring

yaoyinnan opened the pull request:
Support network IO traffic monitoring. Including IO bytes and kps of Redis requests and master-slave replication.

Including the following indicators:
```
total_net_input_bytes: The total number of bytes read from the network
total_net_output_bytes: The total number of bytes written to the network
total_net_repl_input_bytes: The total number of bytes read from the network for replication purposes
total_net_repl_output_bytes: The total number of bytes written to the network for replication purposes
instantaneous_input_kbps: The network's read rate per second in KB/sec
instantaneous_output_kbps: The network's write rate per second in KB/sec
instantaneous_input_repl_kbps: The network's read rate per second in KB/sec for replication purposes
instantaneous_output_repl_kbps: The network's write rate per second in KB/sec for replication purposes
```

Execute the `info stats` command to view:

```
total_net_input_bytes:7655
total_net_output_bytes:84932
total_net_repl_input_bytes:5216
total_net_repl_output_bytes:4013
instantaneous_input_kbps:0.0751953
instantaneous_output_kbps:0.988281
instantaneous_input_repl_kbps:0.03125
instantaneous_output_repl_kbps:0.0185547
```

Collection rules:
- For the Net IO traffic of Redis, it is collected in `GetRequest` and `SendReply` of `RedisConn`.
- For the Net IO traffic of master-slave replication, it is collected in `GetRequest` and `SendReply` of `PbConn`.
- For Net IO rate, the average per second is calculated by recording the value every 0.1s. Use an `instantaneous_metric` thread is started to collect instantaneous data every 0.1s.

Shown in Grafana:

![network](https://github.com/OpenAtomFoundation/pika/assets/35447132/a950ef45-e901-4a7a-a1d0-0e328431e78b)

Fixes: #1732

yaoyinnan commented:
I'm not sure if monotonic_time is doing the right thing, it has to be compatible with x86 and arm of linux and darwin, and each OS and architecture may have different underlying acquisition time. Please take a look here.

luky116 commented on include/pika_instant.h:
```diff
@@ -0,0 +1,41 @@
+
+/* The following two are used to track instantaneous metrics, like
+* number of operations per second, network traffic. */
+struct inst_metric{
```
是否用大小写

yaoyinnan replied:
是的，它应该大写，修改了。

chejinge commented on include/pika_instant.h:
```diff
@@ -0,0 +1,41 @@
+#include <string>
+#include <unordered_map>
+
+#define STATS_METRIC_SAMPLES 16     /* Number of samples per metric. */
```
define改为const

yaoyinnan replied:
Changed to `inline constexpr size_t`.

chejinge commented on include/pika_instant.h:
```diff
@@ -0,0 +1,41 @@
+inline const std::string STATS_METRIC_NET_INPUT_REPLICATION = "stats_metric_net_input_replication";
+inline const std::string STATS_METRIC_NET_OUTPUT_REPLICATION = "stats_metric_net_output_replication";
+
+#define run_with_period(_ms_) if (((_ms_) <= 1000/server.hz) || !(server.cronloops%((_ms_)/(1000/server.hz))))
```
这里用lamda表达式

yaoyinnan replied:
This function is actually not used and has been deleted. The timing execution is modified to start a thread at `pika_server::Start()` to collect instantaneous indicators at regular intervals.

chejinge commented on include/pika_server.h:
```diff
@@ -306,6 +307,19 @@ class PikaServer : public pstd::noncopyable {
+  uint64_t NetInputBytes();
+  uint64_t NetOutputBytes();
+  uint64_t NetReplInputBytes();
+  uint64_t NetReplOutputBytes();
```
uint64_t统一改为size_t

yaoyinnan replied:
edited.

chejinge commented on src/net/include/net_stats.h:
```diff
@@ -0,0 +1,36 @@
+  std::atomic<uint64_t> stat_net_input_bytes = {0}; /* Bytes read from network. */
+  std::atomic<uint64_t> stat_net_output_bytes = {0}; /* Bytes written to network. */
+  std::atomic<uint64_t> stat_net_repl_input_bytes = {0}; /* Bytes read during replication, added to stat_net_input_bytes in 'info'. */
+  std::atomic<uint64_t> stat_net_repl_output_bytes = {0}; /* Bytes written during replication, added to stat_net_output_bytes in 'info'. */
```
= {}二选一，不要两个都写

yaoyinnan replied:
edited.

chejinge commented on src/net/src/net_stats.cc:
```diff
@@ -0,0 +1,30 @@
+
+namespace net {
+
+uint64_t NetworkStatistic::NetInputBytes() { return stat_net_input_bytes.load(std::memory_order_relaxed); }
```
return 和return的内容放在同一行

yaoyinnan replied:
edited.

chejinge commented on src/net/src/net_stats.cc:
```diff
@@ -0,0 +1,30 @@
+
+uint64_t NetworkStatistic::NetReplOutputBytes() { return stat_net_repl_output_bytes.load(std::memory_order_relaxed); }
+
+void NetworkStatistic::IncrInputBytes(uint64_t bytes) { stat_net_input_bytes.fetch_add(bytes, std::memory_order_relaxed); }
```
函数体放在下一行

yaoyinnan replied:
edited.

chejinge commented on src/pika_instant.cc:
```diff
@@ -0,0 +1,39 @@
+  if (inst_metrics_[metric].last_sample_base > 0) {
+    uint64_t base = current_base - inst_metrics_[metric].last_sample_base;
+    uint64_t value = current_value - inst_metrics_[metric].last_sample_value;
+    uint64_t avg = base > 0 ? (value * factor / base) : 0;
```
uint64_t可以统一改成size_t

yaoyinnan replied:
edited.

chejinge commented on src/pika_instant.cc:
```diff
@@ -0,0 +1,39 @@
+
+  for (j = 0; j < STATS_METRIC_SAMPLES; j++)
+    sum += inst_metrics_[metric].samples[j];
+  return sum / STATS_METRIC_SAMPLES;
```
return和上面一行对齐

yaoyinnan replied:
Here return is not included in the for loop.

chejinge commented on src/pika_server.cc:
```diff
@@ -1446,6 +1486,19 @@ void PikaServer::AutoKeepAliveRSync() {
 
+void PikaServer::AutoInstantaneousMetric() {
+  monotime current_time = getMonotonicUs();
+  uint64_t factor = 1000000; // us
```
size_t

yaoyinnan replied:
edited.

luky116 commented on include/pika_server.h:
```diff
@@ -474,6 +488,12 @@ class PikaServer : public pstd::noncopyable {
+ /*
+  * Instantaneous Metric used
+  */
+ Instant instant_;
```
1、格式缩进对齐
2、考虑下是否使用 std::unique_ptr

yaoyinnan replied:
是的

luky116 commented on src/pika_server.cc:
```diff
@@ -192,6 +206,8 @@ void PikaServer::Start() {
   }
   LOG(INFO) << "Goodbye...";
+
+  instantaneous_metric.join();
```
为啥是在 pika exit 的时候进行 join 呢？

yaoyinnan replied:
我想了想，这里不需要 join，instant_metric是后台线程，不需要阻塞主线程。我去掉了。

luky116 commented on src/net/src/redis_conn.cc:
```diff
@@ -87,6 +87,7 @@ ReadStatus RedisConn::GetRequest() {
   }
 
   nread = read(fd(), rbuf_ + next_read_pos, remain);
+  g_network_statistic->IncrInputBytes(nread);
```
一个是在 redis_con 监控，一个是在 pb_conn 监控，命名方式可以考虑统一下：
IncrRedisInputBytes
IncrRedisOutputBytes
和
IncrReplInputBytes
IncrReplOutputBytes

yaoyinnan replied:
是的，这样区分比较好，已经修改了。

luky116 commented on src/pika_admin.cc:
```diff
@@ -840,6 +840,17 @@ void InfoCmd::InfoStats(std::string& info) {
+  tmp_stream << "total_net_output_bytes:" << g_pika_server->NetOutputBytes() + g_pika_server->NetReplOutputBytes() << "\r\n";
+  tmp_stream << "total_net_repl_input_bytes:" << g_pika_server->NetReplInputBytes() << "\r\n";
+  tmp_stream << "total_net_repl_output_bytes:" << g_pika_server->NetReplOutputBytes() << "\r\n";
+  tmp_stream << "instantaneous_input_kbps:" << g_pika_server->InstantaneousInputKbps() << "\r\n";
```
是否要说明是 pika 命令相关的瞬时流量？

yaoyinnan replied:
这里instantaneous_input_kbps所代表的是Redis 命令和Repl的瞬时流量之和。当部署单节点时，total_net_output_bytes和instantaneous_input_kbps只代表Redis 命令的流量。当部署主从集群的时候，只需要通过差值就可以得到 Redis 命令的流量。

yaoyinnan committed:
feat: support network IO traffic monitoring

Support network IO traffic monitoring. Including IO bytes and kps of Redis requests and master-slave replication.

Fixes: #1732

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

yaoyinnan committed:
feat: support network IO traffic monitoring

Support network IO traffic monitoring. Including IO bytes and kps of Redis requests and master-slave replication.

Fixes: #1732

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

yaoyinnan committed:
feat: support network IO traffic monitoring

Support network IO traffic monitoring. Including IO bytes and kps of Redis requests and master-slave replication.

Fixes: #1732

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

yaoyinnan committed:
fix: fix exporter makefile

fix exporter makefile

Fixes: #1732

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>
